<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [36]:
import pandas as pd
import psycopg2

In [37]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [38]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [ ]:
# текст запроса
query_3_1 = f'''
SELECT
  COUNT(DISTINCT id)
FROM
  vacancies
'''

In [ ]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

/tmp/ipython-input-4045908245.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection)


,count
0,49197


2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [ ]:
# текст запроса
query_3_2 = f'''
SELECT
  COUNT(DISTINCT id)
FROM
  employers
'''

In [ ]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

/tmp/ipython-input-533540996.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


,count
0,23501


3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [ ]:
# текст запроса
query_3_3 = f'''
SELECT
  COUNT(DISTINCT id)
FROM
  areas
'''

In [ ]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

/tmp/ipython-input-1395976625.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


,count
0,1362


4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [ ]:
# текст запроса
query_3_4 = f'''
SELECT
  COUNT(DISTINCT id)
FROM
  industries
'''

In [ ]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

/tmp/ipython-input-854197215.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


,count
0,294


***

In [ ]:
# выводы по предварительному анализу данных

1. В базе содержится 49 197 вакансий, что является значительным объемом для рынка.
2. Зарегистрировано 23 501 работодатель. Соотношение вакансий к работодателям составляет примерно 2.1:1, что говорит о том, что в среднем каждый работодатель размещает чуть более 2 вакансий.
3. База охватывает 1 362 региона (города, области, районы). Это свидетельствует о широком географическом охвате платформы.
4. Представлено 294 отрасли/сферы деятельности, что указывает на разнообразие профессиональных областей в базе.

## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
query_4_1 = f'''
SELECT
  a.name,
  COUNT(DISTINCT v.id) AS cnt
FROM
  vacancies AS v
INNER JOIN areas AS a ON v.area_id = a.id
GROUP BY a.name
ORDER BY COUNT(DISTINCT v.id) DESC
LIMIT 5
'''

In [ ]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

/tmp/ipython-input-620650199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


,name,cnt
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = f'''
SELECT
  COUNT(DISTINCT CASE WHEN (salary_from IS NOT NULL) OR (salary_to IS NOT NULL) THEN id ELSE NULL END)
FROM
  vacancies
'''

In [ ]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

/tmp/ipython-input-3746829199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


,count
0,24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [ ]:
# текст запроса
query_4_3 = f'''
SELECT
  ROUND(AVG(CASE WHEN salary_from IS NOT NULL THEN salary_from END), 0) AS mean_low,
  ROUND(AVG(CASE WHEN salary_to IS NOT NULL THEN salary_to END), 0) AS mean_high
FROM
  vacancies
'''

In [ ]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

/tmp/ipython-input-3595259731.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


,mean_low,mean_high
0,71065.0,110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query_4_4 = f'''
SELECT
  schedule,
  employment,
  COUNT(DISTINCT id) AS cnt
FROM
  vacancies
WHERE schedule IS NOT NULL
  AND employment IS NOT NULL
GROUP BY schedule, employment
ORDER BY COUNT(DISTINCT id) DESC
'''

In [ ]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

/tmp/ipython-input-2028972612.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


,schedule,employment,cnt
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [ ]:
# текст запроса
query_4_5 = f'''
SELECT
  experience,
  COUNT(DISTINCT id) AS cnt
FROM
  vacancies
WHERE experience IS NOT NULL
GROUP BY experience
ORDER BY COUNT(DISTINCT id) ASC
'''

In [ ]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

/tmp/ipython-input-210463167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


,experience,cnt
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


***

In [ ]:
# выводы по детальному анализу вакансий

1. Большая часть вакансий располагается в городах-столицах, либо городах-миллионниках.
2. 48.93% вообще упоминают зарплату (по верхней или нижней границе, или по обоим), остальная часть скрывает данную информацию.
3. Примерный разброс зарплаты идет от 70 тыс до 100 тыс, что является неплохой зарплатой в большинстве регионов, не считая столиц.
4. 71.89% вакансий предполагают полную занятость с полной работой.
5. 53.16% вакансий нужен опыт от 1 до 3х лет, и 29.5% требуют опыт от 3 до 6 лет.

## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# текст запроса
query_5_1 = f'''
WITH ranked_employers AS (
  SELECT
    ROW_NUMBER() OVER (ORDER BY COUNT(DISTINCT v.id) DESC) AS position,
    e.name AS employer_name,
    COUNT(DISTINCT v.id) AS cnt
  FROM
    vacancies AS v
  INNER JOIN employers AS e ON v.employer_id = e.id
  GROUP BY e.name
)
SELECT
  position,
  employer_name,
  cnt
FROM ranked_employers
WHERE position IN (1, 5)
ORDER BY position;
'''

In [ ]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

/tmp/ipython-input-1147638897.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


,position,employer_name,cnt
0,1,Яндекс,1933
1,5,Газпром нефть,331


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
query_5_2 = f'''
WITH area_vac AS (
  SELECT
    a.id,
    a.name AS area_name,
    COUNT(v.id) AS vac_cnt
  FROM
    areas AS a
  LEFT JOIN vacancies AS v ON a.id = v.area_id
  GROUP BY a.id, a.name
)
SELECT
  av.id,
  av.area_name,
  av.vac_cnt,
  a.empl_cnt
FROM area_vac AS av
JOIN (
  SELECT
    a.id,
    COUNT(e.id) AS empl_cnt
  FROM areas AS a
  LEFT JOIN employers AS e ON a.id = e.area
  GROUP BY a.id
) AS a ON av.id = a.id
WHERE vac_cnt = 0
ORDER BY empl_cnt DESC
'''

In [ ]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

/tmp/ipython-input-2306256317.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


,id,area_name,vac_cnt,empl_cnt
0,113,Россия,0,410
1,40,Казахстан,0,207
2,2019,Московская область,0,75
3,1438,Краснодарский край,0,19
4,1530,Ростовская область,0,18
...,...,...,...,...
588,5175,Лиман (Астраханская область),0,0
589,4659,Хоринск,0,0
590,1802,Старица,0,0
591,3766,Ессентукская,0,0


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query_5_3 = f'''
SELECT
  e.id AS empl_id,
  e.name AS employer_name,
  COUNT(DISTINCT v.area_id) AS area_cnt
FROM employers AS e
INNER JOIN vacancies AS v ON e.id = v.employer_id
GROUP BY e.id, e.name
ORDER BY COUNT(DISTINCT v.area_id) DESC
'''

In [ ]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

/tmp/ipython-input-133400711.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


,empl_id,employer_name,area_cnt
0,1740,Яндекс,181
1,2748,Ростелеком,152
2,5724811,Спецремонт,116
3,5130287,Поляков Денис Иванович,88
4,3682876,ООО ЕФИН,71
...,...,...,...
14901,810278,НПП Авиатрон,1
14902,810313,Центр дистанционных торгов,1
14903,810551,Городские Телекоммуникационные Системы,1
14904,810688,"Введенский, Отель",1


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [ ]:
# текст запроса
query_5_4 = f'''
WITH employers_without_ind AS (
  SELECT
    e.id AS empl_id,
    e.name AS employer_name,
    COUNT(ei.industry_id) AS ind_cnt
  FROM employers AS e
  LEFT JOIN employers_industries AS ei ON e.id = ei.employer_id
  GROUP BY e.id, e.name
  HAVING COUNT(ei.industry_id) = 0
)
SELECT
  COUNT(*)
FROM
  employers_without_ind
'''

In [ ]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

/tmp/ipython-input-1045119256.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


,count
0,8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [ ]:
# текст запроса
query_5_5 = f'''
WITH ranked_empl AS (
  SELECT
    ROW_NUMBER() OVER (ORDER BY employer_name) AS rank,
    employer_name,
    ind_cnt
  FROM (
    SELECT
      e.id AS empl_id,
      e.name AS employer_name,
      COUNT(ei.industry_id) AS ind_cnt
    FROM employers AS e
    LEFT JOIN employers_industries AS ei ON e.id = ei.employer_id
    GROUP BY e.id, e.name
    HAVING COUNT(ei.industry_id) = 4
  ) AS a
)
SELECT
  employer_name,
  ind_cnt
FROM
  ranked_empl
WHERE rank = 3
'''

In [ ]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

/tmp/ipython-input-4025176131.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


,employer_name,ind_cnt
0,2ГИС,4


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [ ]:
# текст запроса
query_5_6 = f'''
SELECT
  COUNT(*)
FROM employers AS e
JOIN employers_industries AS ei ON e.id = ei.employer_id
JOIN industries AS i ON i.id = ei.industry_id
WHERE i.name LIKE '%Разработка программного обеспечения%'
'''

In [ ]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

/tmp/ipython-input-2713946032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


,count
0,3553


7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [ ]:
import requests
import json
from bs4 import BeautifulSoup

params = {
    'action': 'parse',
    'page': 'Города-миллионеры_России',
    'format': 'json',
    'prop': 'text'
}

api_url = "https://ru.wikipedia.org/w/api.php"

response = requests.get(api_url, params=params,
                       headers={'User-Agent': 'MyBot/1.0'})

if response.status_code == 200:
    data = response.json()
    html_content = data['parse']['text']['*']

    soup = BeautifulSoup(html_content, 'html.parser')
    table = soup.find('table', {'class': 'standard'})

    cities = []
    if table:
        for row in table.find_all('tr')[1:17]:
            columns = row.find_all('td')
            if len(columns) > 1:
                city_name = columns[1].get_text(strip=True)
                cities.append(city_name)

    print("Список городов-миллионников России:")
    for city in cities:
        print(f"- {city}")

    print(f"\nВсего городов: {len(cities)}")

Список городов-миллионников России:
- Москва
- Санкт-Петербург
- Новосибирск
- Екатеринбург
- Казань
- Красноярск
- Нижний Новгород
- Челябинск
- Уфа
- Краснодар
- Самара
- Ростов-на-Дону
- Омск
- Воронеж
- Пермь
- Волгоград

Всего городов: 16


In [ ]:
# текст запроса
values = ",\n        ".join([f"('{city}')" for city in cities])

query_5_7 = f'''
SELECT
  name,
  cnt
FROM (
  SELECT
  0 as "order",
  a.name,
  COUNT(v.id) AS cnt
  FROM areas a
  LEFT JOIN vacancies v ON a.id=v.area_id
  LEFT JOIN employers e ON e.id = v.employer_id
  WHERE a.name in ({values})
    AND e.name = 'Яндекс'
  GROUP BY a.name
  UNION
  SELECT
  1 as "order",
  'Total' as name,
  COUNT(v.id) AS cnt
  FROM areas a
  LEFT JOIN vacancies v ON a.id=v.area_id
  LEFT JOIN employers e ON e.id = v.employer_id
  WHERE a.name in ({values})
    AND e.name = 'Яндекс'
) table_0
ORDER BY "order"
'''

In [ ]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

/tmp/ipython-input-2595257991.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


,name,cnt
0,Волгоград,24
1,Воронеж,32
2,Екатеринбург,39
3,Казань,25
4,Краснодар,30
5,Красноярск,23
6,Москва,54
7,Нижний Новгород,36
8,Новосибирск,35
9,Омск,21


***

In [ ]:
# выводы по анализу работодателей

1. Яндекс является абсолютным лидером на рынке вакансий, что подчеркивает его доминирующую роль в IT-секторе и активный рост.
2. Вакансии сконцентрированы в ограниченном числе регионов (например, Москва и другие крупные города), в то время как многие регионы (например, Казахстан, Московская область) не имеют зарегистрированных вакансий, несмотря на наличие работодателей.
3. Крупные компании, такие как Яндекс и Ростелеком, размещают вакансии в большом количестве регионов.
4. У значительного числа работодателей (8419) не указана сфера деятельности, что усложняет анализ отраслевой структуры рынка труда.
5. Сфера «Разработка программного обеспечения» указана у 3553 работодателей, подтверждая высокий спрос на IT-специалистов. Это согласуется с лидерством Яндекса и наличием компаний вроде 2ГИС в топе по количеству сфер деятельности.
6. Большая часть вакансий создается крупными компаниями в ограниченном числе регионов, что может усиливать региональное неравенство на рынке труда.

## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [ ]:
query_6_1 = f'''
SELECT
COUNT(name)
FROM vacancies
WHERE LOWER(name) LIKE '%data%'
  OR LOWER(name) LIKE '%данн%'
'''

In [ ]:
## результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

/tmp/ipython-input-929020417.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


,count
0,1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [ ]:
query_6_2 = f'''
SELECT
COUNT((name))
FROM vacancies
WHERE (LOWER(name) LIKE '%data scientist%'
  OR LOWER(name) LIKE '%data science%'
  OR LOWER(name) LIKE '%исследователь данных%'
  OR (name LIKE '%ML%' AND NOT name LIKE '%HTML%')
  OR LOWER(name) LIKE '%machine learning%'
  OR LOWER(name) LIKE '%машинн%обучен%')
  AND (LOWER(name) LIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
  )
'''

In [ ]:
## результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

/tmp/ipython-input-2400311998.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


,count
0,51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [50]:
query_6_3 = f'''
SELECT
COUNT(name)
FROM vacancies
WHERE (LOWER(name) LIKE '%data scientist%'
  OR LOWER(name) LIKE '%data science%'
  OR LOWER(name) LIKE '%исследователь данных%'
  OR (name LIKE '%ML%' AND NOT name LIKE '%HTML%')
  OR LOWER(name) LIKE '%machine learning%'
  OR LOWER(name) LIKE '%машинн%обучен%')
  AND (LOWER(key_skills) LIKE '%sql%' OR LOWER(key_skills) LIKE '%postgr%' OR LOWER(key_skills) LIKE '%postr%')
'''

In [51]:
## результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

/tmp/ipython-input-4051982507.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


,count
0,201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [42]:
query_6_4 = f'''
SELECT
COUNT(name)
FROM vacancies
WHERE (LOWER(name) LIKE '%data scientist%'
  OR LOWER(name) LIKE '%data science%'
  OR LOWER(name) LIKE '%исследователь данных%'
  OR (name LIKE '%ML%' AND NOT name LIKE '%HTML%')
  OR LOWER(name) LIKE '%machine learning%'
  OR LOWER(name) LIKE '%машинн%обучен%')
 AND (LOWER(key_skills) LIKE '%python%' OR LOWER(key_skills) LIKE '%pyton%')
'''

In [43]:
## результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

/tmp/ipython-input-2264491117.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


,count
0,351


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [52]:
# текст запроса
query_6_5 = f'''
  SELECT
    ROUND(AVG(skill_count), 2) AS avg_skills_per_vacancy
FROM (
    SELECT
        CASE
            WHEN key_skills IS NULL OR key_skills = '' THEN 0
            ELSE array_length(string_to_array(key_skills, E'\t'), 1)
        END AS skill_count
    FROM vacancies
    WHERE (LOWER(name) LIKE '%data scientist%'
      OR LOWER(name) LIKE '%data science%'
      OR LOWER(name) LIKE '%исследователь данных%'
      OR (name LIKE '%ML%' AND NOT name LIKE '%HTML%')
      OR LOWER(name) LIKE '%machine learning%'
      OR LOWER(name) LIKE '%машинн%обучен%')
) AS skill_counts;
'''

In [53]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

/tmp/ipython-input-4089566367.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


,avg_skills_per_vacancy
0,5.75


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [54]:
# текст запроса
query_6_6 = f'''
  WITH ds_vacancies AS (
    SELECT
        experience,
        salary_from,
        salary_to,
        -- Расчет зарплаты
        CASE
            WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL
                THEN (salary_from + salary_to) / 2.0
            WHEN salary_from IS NOT NULL THEN salary_from
            WHEN salary_to IS NOT NULL THEN salary_to
        END AS calculated_salary
    FROM vacancies
    WHERE
        -- Data Science фильтр
        (LOWER(name) LIKE '%data scientist%'
      OR LOWER(name) LIKE '%data science%'
      OR LOWER(name) LIKE '%исследователь данных%'
      OR (name LIKE '%ML%' AND NOT name LIKE '%HTML%')
      OR LOWER(name) LIKE '%machine learning%'
      OR LOWER(name) LIKE '%машинн%обучен%')
        -- Условие о зарплате
        AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
)
SELECT
    experience,
    COUNT(*) as vacancy_count,
    ROUND(AVG(calculated_salary)) as avg_salary,
    ROUND(MIN(calculated_salary)) as min_salary,
    ROUND(MAX(calculated_salary)) as max_salary
FROM ds_vacancies
GROUP BY experience
ORDER BY
    CASE experience
        WHEN 'Нет опыта' THEN 1
        WHEN 'От 1 года до 3 лет' THEN 2
        WHEN 'От 3 до 6 лет' THEN 3
        WHEN 'Более 6 лет' THEN 4
        ELSE 5
    END;
'''

In [55]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

/tmp/ipython-input-2297777544.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


,experience,vacancy_count,avg_salary,min_salary,max_salary
0,Нет опыта,7,74643.0,37000.0,115000.0
1,От 1 года до 3 лет,28,139675.0,50000.0,275000.0
2,От 3 до 6 лет,27,243115.0,67126.0,450000.0


***

In [ ]:
# выводы по предметному анализу

1. SQL/Postgres — фундаментальный и обязательный навык. Тот факт, что его упоминают в 201 вакансии (и, безусловно, в большинстве DS-вакансий), подтверждает, что работа с реляционными базами данных — неотъемлемая часть работы дата-сайентиста
2. Python — абсолютный стандарт индустрии. Он указан в 351 вакансии из всех, связанных с данными, что делает его самым критичным навыком.
3. Всего 7 вакансий категории "Нет опыта" в выборке DS. Это означает, что новичкам необходимо выделяться: сильным портфолио, проектами, знанием стека технологий.
4. Основной спрос на рынке приходится на специалистов с опытом от 1 до 6 лет (в сумме 55 вакансий). Это ключевой "рабочий" сегмент рынка.
5. Наблюдается явная и существенная положительная корреляция между опытом и уровнем дохода.
6. В среднем около 6 ключевых навыков на вакансию — это высокий показатель. Он подтверждает, что Data Scientist — это мультидисциплинарная роль.

## Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

1. Общая картина рынка: концентрированный и зрелый.

Рынок, представленный на платформе, крупный и активный (~49k вакансий), но характеризуется высокой концентрацией:

Географической: Вакансии сфокусированы в столицах и городах-миллионниках. Многие формально зарегистрированные регионы (особенно целые страны или области) не имеют вакансий, что указывает на серьезное региональное неравенство в предложении цифровых/IT-профессий.

Отраслевой и корпоративной: Доминируют крупные технологические компании (Яндекс, Ростелеком). Сфера IT/Разработка ПО — абсолютный лидер по спросу, что задает тон всему рынку для аналитических и data-ролей.

2. Data Science как профессия: высокие барьеры входа и требовательность.

Ядро стека технологий четко определено: Python и SQL — абсолютный non-negotiable стандарт. Их знание — обязательное условие.

Роль для опытных специалистов: Рынок практически не предлагает входных позиций ("Нет опыта"). Основной спрос (свыше 80% выборки) — на специалистов с опытом от 1 до 6 лет. Это указывает на зрелость области: компании ищут "рабочих лошадок", способных сразу приносить результат.

Мультидисциплинарность подтверждена: Высокое среднее количество навыков на вакансию (~6) подчеркивает, что современный Data Scientist — это гибридная роль (программист, статистик, аналитик, иногда инженер).

Прозрачная ценность опыта: Наблюдается четкая положительная корреляция "опыт — заработная плата", что формирует понятный карьерный трек для специалистов.

3. Институциональные особенности рынка:

Низкая прозрачность по зарплатам: Более половины вакансий не указывают зарплатный диапазон, что усложняет соискателям оценку своего рыночного уровня и требует дополнительного исследования.

Данные часто неполны: Значительная доля работодателей не указала сферу деятельности, что говорит либо о необязательности этого поля, либо о сложности классификации, что "зашумляет" общую картину.

### Дополнительные исследования

**Гипотеза**: Несмотря на лишь 7 вакансий "без опыта", часть вакансий "от 1 года опыта" может быть de facto entry-level, где требование формально.

**Проверка**: Сравнить средние предлагаемые зарплаты и стек технологий для вакансий "Нет опыта" и "1–3 года". Если они статистически неразличимы, это подтвердит гипотезу, что стартовать можно и через портфель проектов, претендуя на junior-позиции.

In [ ]:
# текст запроса
query_7_1 = f'''

'''

In [ ]:
# результат запроса
df_7_1 = pd.read_sql_query(query_7_1, connection)
df_7_1